stm32: gpio: Fix device runtime pm

ceolin opened the pull request:
When CONFIG_PM_DEVICE_RUNTIME is enabled, if a pin is configure as
input after an output pin has already being configured the device is
wrongly suspended.

Fixes #38433

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
stm32: gpio: Fix device runtime pm

When CONFIG_PM_DEVICE_RUNTIME is enabled, if a pin is configure as
input after an output pin has already being configured the device is
wrongly suspended.

Fixes #38433

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

aurel32 commented:
Thanks a lot for the PR, I confirm that it fixes the issue I reported. About the fix itself, I wonder if it could be a bit fragile, as it is not thread nor async safe. The latter case might be problematic, although it works in my original case (STM32L4 + SX1276), as `pm_device_put_async()` is used at the end. But maybe I am just wrong here.

Another alternative I have just tested is to just use the PM reference counting of the device (`dev->pm->usage`), as there is a fast path if the device is already in the correct state. This is basically the following patch:

```
diff --git a/drivers/gpio/gpio_stm32.c b/drivers/gpio/gpio_stm32.c
index 056594818a..40e53936dc 100644
--- a/drivers/gpio/gpio_stm32.c
+++ b/drivers/gpio/gpio_stm32.c
@@ -464,15 +464,10 @@ static int gpio_stm32_config(const struct device *dev,
 	}
 
 #ifdef CONFIG_PM_DEVICE_RUNTIME
-	enum pm_device_state state;
-
-	(void)pm_device_state_get(dev, &state);
 	/* Enable device clock before configuration (requires bank writes) */
-	if (state != PM_DEVICE_STATE_ACTIVE) {
-		err = pm_device_get(dev);
-		if (err < 0) {
-			return err;
-		}
+	err = pm_device_get(dev);
+	if (err < 0) {
+		return err;
 	}
 #endif /* CONFIG_PM_DEVICE_RUNTIME */
```

However I don't know the PM subsystem enough to decide if one solution is better than the other.

ceolin commented:
> Thanks a lot for the PR, I confirm that it fixes the issue I reported. About the fix itself, I wonder if it could be a bit fragile, as it is not thread nor async safe. The latter case might be problematic, although it works in my original case (STM32L4 + SX1276), as `pm_device_put_async()` is used at the end. But maybe I am just wrong here.
> 
> Another alternative I have just tested is to just use the PM reference counting of the device (`dev->pm->usage`), as there is a fast path if the device is already in the correct state. This is basically the following patch:
> 
> ```
> diff --git a/drivers/gpio/gpio_stm32.c b/drivers/gpio/gpio_stm32.c
> index 056594818a..40e53936dc 100644
> --- a/drivers/gpio/gpio_stm32.c
> +++ b/drivers/gpio/gpio_stm32.c
> @@ -464,15 +464,10 @@ static int gpio_stm32_config(const struct device *dev,
>  	}
>  
>  #ifdef CONFIG_PM_DEVICE_RUNTIME
> -	enum pm_device_state state;
> -
> -	(void)pm_device_state_get(dev, &state);
>  	/* Enable device clock before configuration (requires bank writes) */
> -	if (state != PM_DEVICE_STATE_ACTIVE) {
> -		err = pm_device_get(dev);
> -		if (err < 0) {
> -			return err;
> -		}
> +	err = pm_device_get(dev);
> +	if (err < 0) {
> +		return err;
>  	}
>  #endif /* CONFIG_PM_DEVICE_RUNTIME */
> ```
> 
> However I don't know the PM subsystem enough to decide if one solution is better than the other.

Yep, you are right about the async problem. I implemented this way because I'm not familiar with the driver and wanted to keep the same behavior.

Though, probably the best thing to do is avoid this optimization and and always call `pm_device_get()`. Will change the patch.

ceolin commented:
> Yep, you are right about the async problem. I implemented this way because I'm not familiar with the driver and wanted to keep the same behavior.
> 
> Though, probably the best thing to do is avoid this optimization and and always call `pm_device_get()`. Will change the patch.

I've changed the patch, though I'm concern about it have unbalanced get/put calls leading the device to be always active. Though this is a *possible* problem that exists independent of this change.

erwango commented:
> I've changed the patch, though I'm concern about it have unbalanced get/put calls leading the device to be always active. Though this is a possible problem that exists independent of this change.

As discussed during the implementation of this piece of code, at the end this is GPIO user's responsibility.
User should know if a pin is required and release it when no more required to switch off the port clock.
Driver should only reflect user's needs (but it's still possible that issues lie in that matter).

aurel32 reviewed (approved):
Thanks, that looks good to me and fixes the issue I reported.
